### PR TITLE
fix #1578 and some improvements

### LINKED ISSUE #1578: failed to train with remote mode
hello, everyone,
      I used the experiment examples/trials/weight_sharing/ga_squad/config_remote.yml, and I modified the machineList with my own machine, but the experiment was waiting and the log is below. Is there any suggestion?
```
[9/29/2019, 4:40:41 PM] WARNING [ 'Scheduler: trialJob id iwzEY, no machine can be scheduled, return TMP_NO_AVAILABLE_GPU ' ]
[9/29/2019, 4:40:41 PM] INFO [ 'Right now no available GPU can be allocated for trial iwzEY, will try to schedule later' ]
[9/29/2019, 4:40:44 PM] WARNING [ 'Scheduler: trialJob id iwzEY, no machine can be scheduled, return TMP_NO_AVAILABLE_GPU ' ]
[9/29/2019, 4:40:44 PM] INFO [ 'Right now no available GPU can be allocated for trial iwzEY, will try to schedule later' ]
[9/29/2019, 4:40:47 PM] WARNING [ 'Scheduler: trialJob id iwzEY, no machine can be scheduled, return TMP_NO_AVAILABLE_GPU ' ]
[9/29/2019, 4:40:47 PM] INFO [ 'Right now no available GPU can be allocated for trial iwzEY, will try to schedule later' ]
[9/29/2019, 4:40:50 PM] WARNING [ 'Scheduler: trialJob id iwzEY, no machine can be scheduled, return TMP_NO_AVAILABLE_GPU ' ]
[9/29/2019, 4:40:50 PM] INFO [ 'Right now no available GPU can be allocated for trial iwzEY, will try to schedule later' ]
[9/29/2019, 4:40:53 PM] WARNING [ 'Scheduler: trialJob id iwzEY, no machine can be scheduled, return TMP_NO_AVAILABLE_GPU ' ]
[9/29/2019, 4:40:53 PM] INFO [ 'Right now no available GPU can be allocated for trial iwzEY, will try to schedule later' ]
[9/29/2019, 4:40:56 PM] WARNING [ 'Scheduler: trialJob id iwzEY, no machine can be scheduled, return TMP_NO_AVAILABLE_GPU ' ]
[9/29/2019, 4:40:56 PM] INFO [ 'Right now no available GPU can be allocated for trial iwzEY, will try to schedule later' ]
[9/29/2019, 4:40:59 PM] WARNING [ 'Scheduler: trialJob id iwzEY, no machine can be scheduled, return TMP_NO_AVAILABLE_GPU ' ]
[9/29/2019, 4:40:59 PM] INFO [ 'Right now no available GPU can be allocated for trial iwzEY, will try to schedule later' ]
[9/29/2019, 4:41:02 PM] WARNING [ 'Scheduler: trialJob id iwzEY, no machine can be scheduled, return TMP_NO_AVAILABLE_GPU ' ]
[9/29/2019, 4:41:02 PM] INFO [ 'Right now no available GPU can be allocated for trial iwzEY, will try to schedule later' ]
[9/29/2019, 4:41:05 PM] WARNING [ 'Scheduler: trialJob id iwzEY, no machine can be scheduled, return TMP_NO_AVAILABLE_GPU ' ]
[9/29/2019, 4:41:05 PM] INFO [ 'Right now no available GPU can be allocated for trial iwzEY, will try to schedule later' ]
[9/29/2019, 4:41:08 PM] WARNING [ 'Scheduler: trialJob id iwzEY, no machine can be scheduled, return TMP_NO_AVAILABLE_GPU ' ]
[9/29/2019, 4:41:08 PM] INFO [ 'Right now no available GPU can be allocated for trial iwzEY, will try to schedule later' ]
[9/29/2019, 4:41:11 PM] WARNING [ 'Scheduler: trialJob id iwzEY, no machine can be scheduled, return TMP_NO_AVAILABLE_GPU ' ]
[9/29/2019, 4:41:11 PM] INFO [ 'Right now no available GPU can be allocated for trial iwzEY, will try to schedule later' ]
[9/29/2019, 4:41:14 PM] WARNING [ 'Scheduler: trialJob id iwzEY, no machine can be scheduled, return TMP_NO_AVAILABLE_GPU ' ]
[9/29/2019, 4:41:14 PM] INFO [ 'Right now no available GPU can be allocated for trial iwzEY, will try to schedule later' ]
[9/29/2019, 4:41:17 PM] WARNING [ 'Scheduler: trialJob id iwzEY, no machine can be scheduled, return TMP_NO_AVAILABLE_GPU ' ]
[9/29/2019, 4:41:17 PM] INFO [ 'Right now no available GPU can be allocated for trial iwzEY, will try to schedule later' ]
[9/29/2019, 4:41:20 PM] WARNING [ 'Scheduler: trialJob id iwzEY, no machine can be scheduled, return TMP_NO_AVAILABLE_GPU ' ]
[9/29/2019, 4:41:20 PM] INFO [ 'Right now no available GPU can be allocated for trial iwzEY, will try to schedule later' ]
[9/29/2019, 4:41:23 PM] WARNING [ 'Scheduler: trialJob id iwzEY, no machine can be scheduled, return TMP_NO_AVAILABLE_GPU ' ]
[9/29/2019, 4:41:23 PM] INFO [ 'Right now no available GPU can be allocated for trial iwzEY, will try to schedule later' ]
[9/29/2019, 4:41:26 PM] WARNING [ 'Scheduler: trialJob id iwzEY, no machine can be scheduled, return TMP_NO_AVAILABLE_GPU ' ]
[9/29/2019, 4:41:26 PM] INFO [ 'Right now no available GPU can be allocated for trial iwzEY, will try to schedule later' ]
[9/29/2019, 4:41:29 PM] WARNING [ 'Scheduler: trialJob id iwzEY, no machine can be scheduled, return TMP_NO_AVAILABLE_GPU ' ]
[9/29/2019, 4:41:29 PM] INFO [ 'Right now no available GPU can be allocated for trial iwzEY, will try to schedule later' ]
[9/29/2019, 4:41:32 PM] WARNING [ 'Scheduler: trialJob id iwzEY, no machine can be scheduled, return TMP_NO_AVAILABLE_GPU ' ]
[9/29/2019, 4:41:32 PM] INFO [ 'Right now no available GPU can be allocated for trial iwzEY, will try to schedule later' ]
[9/29/2019, 4:41:35 PM] WARNING [ 'Scheduler: trialJob id iwzEY, no machine can be scheduled, return TMP_NO_AVAILABLE_GPU ' ]
```

### PR DESCRIPTION
**Pull request by squirrelsc**
Add shell support for ssh connection, so that remote script can be started with user environment.

Minor fixes,

1. Fix gpu_metrics_collector to support pyenv. As pyenv will create one more process, so that original pgrep code always got extra processes, and cannot start gpu_metrics_collector.
2. Fix NASUI failure on dev-install-node-modules, to create subfolder every time.
3. Add .gitignore items for dev-install.
4. Add node --watch for nni_manager for better dev experience.

dev env test passed.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -84,3 +84,8 @@ venv.bak/
 
 # In case you place source code in ~/nni/
 /experiments
+src/sdk/pynni/nni/nni
```
Comment (SparkSnail) [inline .gitignore:87]: why add these folders?

File: src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts [hunk 2]
```diff
@@ -442,7 +442,7 @@ class RemoteMachineTrainingService implements TrainingService {
 
         //Begin to execute gpu_metrics_collection scripts
         const script = getGpuMetricsCollectorBashScriptContent(remoteGpuScriptCollectorDir);
-        SSHClientUtility.remoteExeCommand(`bash -c '${script}'`, conn);
+        SSHClientUtility.remoteExeCommand(`bash -c '${script}'`, conn, true);
```
Comment (SparkSnail) [inline src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts:445]: what is the purpose of `true` here?

File: src/nni_manager/training_service/remote_machine/sshClientUtility.ts [hunk 3]
```diff
@@ -58,38 +58,56 @@ export namespace SSHClientUtility {
...
      */
-    export function remoteExeCommand(command: string, client: Client): Promise<RemoteCommandResult> {
+    export function remoteExeCommand(command: string, client: Client, useShell: boolean = false): Promise<RemoteCommandResult> {
         const log: Logger = getLogger();
         log.debug(`remoteExeCommand: command: [${command}]`);
         const deferred: Deferred<RemoteCommandResult> = new Deferred<RemoteCommandResult>();
         let stdout: string = '';
         let stderr: string = '';
         let exitCode: number;
+        let captureStdOut = false;
```
Comment (SparkSnail) [inline src/nni_manager/training_service/remote_machine/sshClientUtility.ts:68]: what is the purpose of `captureStdout`?

File: tools/nni_gpu_tool/gpu_metrics_collector.py [hunk 4]
```diff
@@ -10,27 +10,32 @@
...
             'wmic process where "CommandLine like \'%nni_gpu_tool.gpu_metrics_collector%\' and name like \'%python%\'" get processId')
         pidList = pgrep_output.decode("utf-8").strip().split()
-        pidList.pop(0) # remove the key word 'ProcessId'
+        pidList.pop(0)  # remove the key word 'ProcessId'
         pidList = list(map(int, pidList))
         pidList.remove(os.getpid())
         return not pidList
     else:
-        pgrep_output = subprocess.check_output('pgrep -fxu "$(whoami)" \'python3 -m nni_gpu_tool.gpu_metrics_collector\'', shell=True)
+        pgrep_output = subprocess.check_output(
```
Comment (ultmaster) [inline tools/nni_gpu_tool/gpu_metrics_collector.py:24]: Our length limit is 140.

File: tools/nni_gpu_tool/gpu_metrics_collector.py [hunk 5]
```diff
@@ -10,27 +10,32 @@
...
+            'pgrep -afu "$(whoami)" \'python3 -m nni_gpu_tool.gpu_metrics_collector\'', shell=True)
```
Comment (SparkSnail) [inline tools/nni_gpu_tool/gpu_metrics_collector.py:25]: why use `-a` instead of `-x`? 
`

File: .gitignore (hunk 1 above)
Comment (ultmaster) [inline .gitignore:87]: I think these are wrong soft links. We should investigate in why they are created.

Comment (squirrelsc) [reply .gitignore:87]: They are links, and created by dev-install.

File: src/nni_manager/training_service/remote_machine/sshClientUtility.ts (hunk 3 above)
Comment (squirrelsc) [reply src/nni_manager/training_service/remote_machine/sshClientUtility.ts:68]: When the shell is using user's interactive environment, there are startup outputs, like what we see when login a SSH shell. This kind of outputs won't be captured, and output.

File: tools/nni_gpu_tool/gpu_metrics_collector.py (hunk 5 above)
Comment (squirrelsc) [reply tools/nni_gpu_tool/gpu_metrics_collector.py:25]: -x means exact match and -a means output command text also. It works with default Python, but if user uses multiple Python tools, like pyenv, the command line won't be exact match like this. So use a more complex filter to handle this situation.

**Comment by chicm-ms** [comment]
suggest to merge into a dev branch to run integration-test-remote  and integration-test-remote-windows pipelines first

**Comment by squirrelsc** [comment]
> suggest to merge into a dev branch to run integration-test-remote and integration-test-remote-windows pipelines first

@chicm-ms 
what's process to create a dev branch, and how to enable those tests on dev branch?

File: Makefile [hunk 6]
```diff
@@ -211,6 +211,7 @@ dev-install-node-modules:
 	sed -ie 's/$(NNI_VERSION_TEMPLATE)/$(NNI_VERSION_VALUE)/' $(NNI_PKG_FOLDER)/package.json
 	ln -sf ${PWD}/src/nni_manager/node_modules $(NNI_PKG_FOLDER)/node_modules
 	ln -sf ${PWD}/src/webui/build $(NNI_PKG_FOLDER)/static
+	mkdir -p $(NASUI_PKG_FOLDER)
 	ln -sf ${PWD}/src/nasui/build $(NASUI_PKG_FOLDER)/build
 	ln -sf ${PWD}/src/nasui/server.js $(NASUI_PKG_FOLDER)/server.js
```
Comment (leckie-chn) [inline Makefile:216]: needs to remove this line, or will give 
```
# Installing NNI Package 
rm -rf /home/yann/.local/nni
ln -sf /home/yann/code/nni/src/nni_manager/dist /home/yann/.local/nni
cp src/nni_manager/package.json /home/yann/.local/nni
sed -ie 's/999.0.0-developing/v0.2-1270-g06bc964/' /home/yann/.local/nni/package.json
ln -sf /home/yann/code/nni/src/nni_manager/node_modules /home/yann/.local/nni/node_modules
ln -sf /home/yann/code/nni/src/webui/build /home/yann/.local/nni/static
mkdir -p /home/yann/.local/nni/nasui
ln -sf /home/yann/code/nni/src/nasui/build /home/yann/.local/nni/nasui/build
ln -sf /home/yann/code/nni/src/nasui/server.js /home/yann/.local/nni/nasui/server.js
ln: '/home/yann/code/nni/src/nasui/server.js' and '/home/yann/.local/nni/nasui/server.js' are the same file
Makefile:207: recipe for target 'dev-install-node-modules' failed
make: *** [dev-install-node-modules] Error 1
```

Comment (squirrelsc) [reply Makefile:216]: Didn't hit this issue, and it's not changed this time.

**Comment by squirrelsc** [comment]
Create #2370 as a dev branch. please continue review there.

**Pull request closed without merging**